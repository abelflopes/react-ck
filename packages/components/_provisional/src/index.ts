export * from "./image";

export * from "./avatar";

export * from "./dropdown";

export * from "./dock";

export * from "./breadcrumbs";

export * from "./scrollable-container";

export * from "./position-engine";

export * from "./top-bar";

export * from "./snackbar";

export * from "./overlay-spinner";

export * from "./menu";

export * from "./tooltip";

export * from "./popover";

export * from "./chat";

export * from "./file-uploader";

export * from "./attachment";

export * from "./snippet";
