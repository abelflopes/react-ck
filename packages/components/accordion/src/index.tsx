import React from "react";
import { List, type ListProps } from "@react-ck/list";
import { Collapse } from "@react-ck/collapse";
import styles from "./styles/index.module.scss";

export interface AccordionItem {
  header: NonNullable<React.ReactNode>;
  children: NonNullable<React.ReactNode>;
}

export interface AccordionProps extends Omit<ListProps, "items"> {
  items: AccordionItem[];
}

/**
 * An accordion is a vertically stacked list of interactive items. Each item can be "collapsed" with only a summary visible or “expanded” to show the full content for that item.
 * @param props - {@link AccordionProps}
 * @returns a React element
 */

export const Accordion = ({
  items,
  ...otherProps
}: Readonly<AccordionProps>): React.ReactElement => (
  <List
    {...otherProps}
    items={items.map(({ header, children }) => (
      <Collapse key={String(JSON.stringify(header))} header={header}>
        <div className={styles.content}>{children}</div>
      </Collapse>
    ))}
  />
);
