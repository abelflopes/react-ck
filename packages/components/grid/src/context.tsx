import React, { useContext } from "react";

export interface GridContextProps {
  columnsCount: number;
  registerColumn: () => () => void;
}

export const themeContextDefaults: GridContextProps = {
  columnsCount: 0,
  registerColumn() {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    return () => {};
  },
};

export const GridContext = React.createContext<GridContextProps>(themeContextDefaults);

export const useGridContext = (): GridContextProps => useContext(GridContext);
