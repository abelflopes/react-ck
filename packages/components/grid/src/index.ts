export * from "./Container";

export * from "./Column";
