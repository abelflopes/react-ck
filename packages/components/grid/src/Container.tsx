import React, { useCallback, useMemo, useState } from "react";
import styles from "./styles/container.module.scss";
import classNames from "classnames";
import { GridContext, type GridContextProps } from "./context";

export interface GridContainerProps extends React.HTMLAttributes<HTMLHRElement> {
  /** The spacing between columns in the grid  */
  spacing?: "s" | "m" | "l" | "none";
  /** Whether to allow grid items to wrap to the next line  */
  wrap?: boolean;
}

export const GridContainer = ({
  spacing = "m",
  wrap = true,
  className,
  ...otherProps
}: Readonly<GridContainerProps>): React.ReactElement => {
  const [columnsCount, setColumnsCount] = useState(0);

  const registerColumn = useCallback<GridContextProps["registerColumn"]>(() => {
    setColumnsCount((v) => v + 1);

    return () => {
      setColumnsCount((v) => v - 1);
    };
  }, []);

  const contextValue = useMemo<GridContextProps>(
    () => ({
      columnsCount,
      registerColumn,
    }),
    [columnsCount, registerColumn],
  );

  return (
    <GridContext.Provider value={contextValue}>
      <div
        className={classNames(
          styles.root,
          styles["spacing_" + spacing],
          {
            [`${styles.wrap}`]: wrap,
          },
          className,
        )}
        {...otherProps}
      />
    </GridContext.Provider>
  );
};
