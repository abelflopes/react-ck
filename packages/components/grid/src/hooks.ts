import { useContext } from "react";
import { GridContext, type GridContextProps } from "./context";

export const useGridContext = (): GridContextProps => useContext(GridContext);
