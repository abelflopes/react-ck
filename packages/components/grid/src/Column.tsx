import React, { useEffect, useState } from "react";
import styles from "./styles/column.module.scss";
import classNames from "classnames";
import { useGridContext } from "./hooks";
import { useResponsiveProps, type ResponsiveProps } from "@react-ck/responsive";

interface BaseProps extends React.HTMLAttributes<HTMLHRElement> {
  /** Specifies the size of the column.
   *  - When omitted will adapt to content size
   *  - When set to `auto` fits to the available space */
  size?: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | "auto";
}

export type GridColumnProps = BaseProps & ResponsiveProps<BaseProps>;

export const GridColumn = ({
  responsive,
  ...baseProps
}: Readonly<GridColumnProps>): React.ReactElement => {
  const { size, className, style, ...otherProps } = useResponsiveProps<BaseProps>({
    baseProps,
    responsive,
  });

  const [index, setIndex] = useState<number | undefined>(undefined);
  const { columnsCount, registerColumn } = useGridContext();

  useEffect(() => {
    const { index, deregister } = registerColumn();

    setIndex(index);

    return deregister;
  }, [registerColumn]);

  return (
    <div
      style={{
        ...style,
        ...{ "--size": size, "--columns": columnsCount, "--index": index },
      }}
      className={classNames(
        styles.root,
        {
          [`${styles.size_defined}`]: typeof size === "number",
          [`${styles.size_auto}`]: size === "auto",
        },
        className,
      )}
      {...otherProps}
    />
  );
};
