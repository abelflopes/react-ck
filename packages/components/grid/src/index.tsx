import React, { useCallback, useMemo, useState } from "react";
import styles from "./styles/container.module.scss";
import classNames from "classnames";
import { GridContext, type GridContextProps } from "./context";
import { GridColumn } from "./Column";

interface GridProps extends React.HTMLAttributes<HTMLHRElement> {
  /** The spacing between columns in the grid  */
  spacing?: "s" | "m" | "l" | "none";
  /** Whether to allow grid items to wrap to the next line  */
  wrap?: boolean;
}

/**
 * A container used to build layouts that align to a user-defined system of columns and rows.
 * @param props - {@link DividerProps}
 * @returns a React element
 */

const Grid = ({
  spacing = "m",
  wrap = true,
  className,
  ...otherProps
}: Readonly<GridProps>): React.ReactElement => {
  const [columnsCount, setColumnsCount] = useState(0);

  const registerColumn = useCallback<GridContextProps["registerColumn"]>(() => {
    setColumnsCount((v) => v + 1);

    return () => {
      setColumnsCount((v) => v - 1);
    };
  }, []);

  const contextValue = useMemo<GridContextProps>(
    () => ({
      columnsCount,
      registerColumn,
    }),
    [columnsCount, registerColumn],
  );

  return (
    <GridContext.Provider value={contextValue}>
      <div
        className={classNames(
          styles.root,
          styles[`spacing_${spacing}`],
          {
            [`${styles.wrap}`]: wrap,
          },
          className,
        )}
        {...otherProps}
      />
    </GridContext.Provider>
  );
};

Grid.Column = GridColumn;

export { Grid, type GridProps };
