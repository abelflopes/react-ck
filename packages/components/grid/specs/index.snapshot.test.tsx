import React from "react";
import { GridContainer, GridColumn } from "../src/index";
import renderer from "react-test-renderer";

describe("snapshot GridContainer", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<GridContainer />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("snapshot GridColumn", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<GridColumn />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
