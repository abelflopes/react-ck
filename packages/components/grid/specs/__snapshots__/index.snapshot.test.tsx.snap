// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot GridColumn renders correctly 1`] = `
<div
  className="root size_auto"
  style={
    {
      "--columns": 0,
      "--size": "auto",
    }
  }
/>
`;

exports[`snapshot GridContainer renders correctly 1`] = `
<div
  className="root spacing_m wrap"
/>
`;
